src/main/kotlin/org/rust/toml/RsTomlCompletionProvider.kt

```diff
@@ -0,0 +1,192 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionProvider
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.patterns.ElementPattern
+import com.intellij.patterns.PlatformPatterns
+import com.intellij.patterns.VirtualFilePattern
+import com.intellij.psi.PsiElement
+import com.intellij.util.ProcessingContext
+import org.intellij.lang.annotations.Language
+import org.rust.lang.core.psi.ext.parentOfType
+import org.toml.lang.psi.*
+
+
+class CargoTomlKeysCompletionProvider : CompletionProvider<CompletionParameters>() {
+    private var cachedSchema: TomlSchema? = null
+
+    override fun addCompletions(
+        parameters: CompletionParameters,
+        context: ProcessingContext?,
+        result: CompletionResultSet
+    ) {
+        val schema = cachedSchema ?:
+            TomlSchema(parameters.position.project, EXAMPLE_CARGO_TOML).also { cachedSchema = it }
+
+        val key = parameters.position.parent as? TomlKey ?: return
+        val table = key.topLevelTable ?: return
+        val parent = key.parent
+        val variants = when (parent) {
+            is TomlTableHeader -> {
+                if (key != parent.names.firstOrNull()) return
+                val isArray = when (table) {
+                    is TomlArrayTable -> true
+                    is TomlTable -> false
+                    else -> return
+                }
+                schema.topLevelKeys(isArray)
+            }
+
+            is TomlKeyValue -> {
+                val tableName = (table as? TomlHeaderOwner)?.header?.names?.firstOrNull()?.text
+                    ?: return
+                schema.keysForTable(tableName)
+            }
+
+            else -> return
+        }
+
+        result.addAllElements(variants.map {
+            LookupElementBuilder.create(it)
+        })
+    }
+
+    private val TomlKey.topLevelTable: TomlKeyValueOwner?
+        get() {
+            val table = parentOfType<TomlKeyValueOwner>() ?: return null
+            if (table.parent !is TomlFile) return null
+            return table
+        }
+
+    companion object {
+        val elementPattern: ElementPattern<PsiElement>
+            get() = PlatformPatterns.psiElement()
+                .inVirtualFile(VirtualFilePattern().withName("Cargo.toml"))
+                .withParent(TomlKey::class.java)
+    }
+}
+
+
+
+@Language("TOML")
+private val EXAMPLE_CARGO_TOML = """
```

Comment:
Hmm, this looks like something temporary? Really we want to implement it like "cargo run" completion works? (it also make possible to complete values like paths or crates).

Reply:
I think that for basic "what keys are available in this table" this example approach is the long-term preferable. The toml is from http://doc.crates.io/manifest.html, so it should be easy to update it when new features are implemented in Cargo. I'll ad the manifest.html URL as a comment here. 

I wish we had a similar example with all Cargo command line options, so that the basic ones (without arguments) could be generated and not hard-coded. 

But yeah, we'll need to implement smarter completion here in the future, like suggesting a name of the package in `[dependenices]` section from the crates on crates.io.

Reply:
So may be it's better to move it to a separate file instead of inline string?

Reply:
And what are the downsides to using an inline string? :) 

The upside is that we don't need to do IO to get the text from resources.

Reply:
The IO is just hidden with java class loading =) And anyway it performs once. So I don't think it is something major.

The downsides is that, hmm, we include a large fragment of non-kotlin code to kotiln file. This looks ugly for me, but I can't say more precisely :)

Reply:
I don't feel strongly either way, and I am to lazy to extract this into a separate file :P

Reply:
Okay =)